Migrating CsrfTokenService

Showtim3 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
I can't get this test to pass in the csrfTokenService test file
```it('should error if getTokenAsync is called before initialize', () => {
    expect(csrfTokenService.getTokenAsync)
      .toThrowError('Token needs to be initialized');
  });
```
## Checklist
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [ ] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `python -m scripts.run_frontend_tests`.
- [x] The PR is made from a branch that's **not** called "develop".
- [X] The PR has an appropriate "PROJECT: ..." label (Please add this label for the first-pass review of the PR).
- [X] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [ ] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [ ] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

Showtim3 committed:
upgrading csrfTokenService

Showtim3 committed:
merging develop

jameesjohn commented on core/templates/dev/head/services/CsrfTokenServiceSpec.ts:
```diff
@@ -24,54 +24,39 @@ import $ from 'jquery';
+
+  beforeEach(() => {
+    csrfTokenService = new CsrfTokenService();
+    // spyOn($, 'ajax').and.returnValue($q);
```
@Showtim3 any reason for this?
I think we should still spy on $.ajax since that is what is used to send the ajax request.

vojtechjelinek commented on core/templates/dev/head/services/CsrfTokenService.ts:
```diff
@@ -19,37 +19,45 @@
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
 import $ from 'jquery';
```
Can we try to import only the `ajax` method from jquery?

Showtim3 replied:
I did a google search for that..didn't find anything relevant

kevinlee12 commented:
@Showtim3 some tests are failing and there is a merge conflict, ptal!

Showtim3 committed:
merging develop

Showtim3 committed:
fixing imports

Showtim3 committed:
cant get a test to pass

kevinlee12 commented:
@vojtechjelinek can you advise? thanks!

vojtechjelinek commented on core/templates/dev/head/services/CsrfTokenServiceSpec.ts:
```diff
@@ -50,6 +47,8 @@ describe('Csrf Token Service', function() {
   });
 
   it('should error if getTokenAsync is called before initialize', () => {
+    csrfTokenService.getTokenAsync();
```
Why are you adding this? This will throw `'Token needs to be initialized'`.

Showtim3 replied:
was trying all the possible permutations! Since i can't get it to pass.

Showtim3 committed:
Removing extra function call

Showtim3 committed:
fixing tests

Showtim3 committed:
merging develop

Showtim3 committed:
fixing imports

Showtim3 committed:
ordering imports

Showtim3 committed:
Adding comment

Showtim3 commented:
@Jamesjay4199 @vojtechjelinek @bansalnitish Hey guys, this is now ready for review.
PTAL! Thanks!

Showtim3 committed:
ts-ignore comment

Showtim3 committed:
remvoing whitespace

Showtim3 committed:
fixing imports

seanlip commented on core/templates/dev/head/services/csrf-token.service.spec.ts:
```diff
@@ -21,57 +21,40 @@
+
+    // This ts ignore is because returnValue is expecting a jqXHR<any>
+    // and wherease a Promise<{token: string}> is being provided.
+    // @ts-ignore
```
I noticed this and wanted to ask about it. Shouldn't we find a way to fix it properly rather than using ts-ignore? /cc @bansalnitish @ankita240796

Showtim3 replied:
@seanlip I tried some hacks to solve this but I don't think there is a clean way to do this.  Since the value it expects is a type of Jquery Object.

seanlip replied:
Understood. I still think you should discuss it with the team though. At the very least an issue should be filed explaining all the things that don't work, but ideally a general solution can be found, since this is going to come up again in other files.

ankita240796 replied:
I will look into this once and update here.

bansalnitish replied:
Let's file an issue for this @Showtim3, also please update your PR.

seanlip reviewed (changes requested):
(See inline comment https://github.com/oppia/oppia/pull/7818#discussion_r343340033)

Showtim3 committed:
Mering develop

vojtechjelinek reviewed (approved):
LGTM from code owner's perspective. Thanks!

kevinlee12 commented:
@seanlip @Jamesjay4199 @bansalnitish ptal!

seanlip commented:
My [comment](https://github.com/oppia/oppia/pull/7818#discussion_r344017599) is not addressed. /cc @bansalnitish @ankita240796

Showtim3 committed:
Merging develop

Showtim3 committed:
fixing typescript

jameesjohn reviewed (approved):
LGTM!

Showtim3 committed:
Adding todo for ts-ignore

Showtim3 committed:
Merge branch 'develop' into migration-CsrfTokenService

Showtim3 committed:
merged develop

Showtim3 commented:
@bansalnitish @seanlip Done! Created a todo for removing the ts-ignore https://github.com/oppia/oppia/issues/8035 and added a todo for the same in the code as a comment. PTAL!

seanlip commented:
@Showtim3 I spoke to @bansalnitish about this. The issue needs more detail -- currently there isn't enough information to determine what analysis has been done so far. Please see https://github.com/oppia/oppia/pull/7818#discussion_r344017599 for guidance.

/cc @ankita240796

ankita240796 commented:
Hi @seanlip @bansalnitish @Showtim3, I am sorry for the delay on this. I am a bit busy with the Nov release and release automation stuff and want to wrap it up by Friday. I will start with the Angular migration stuff after that. @bansalnitish can you please guide @Showtim3 meanwhile and get the issue [here](https://github.com/oppia/oppia/pull/7818#discussion_r344017599) resolved. I will try to look into it before Friday if I get time. Really sorry for the delay. Thanks!

bansalnitish reviewed (approved):
LGTM!

Even I'm not sure about dropping that todo -- definitely need some more time and investigation. Can we file a bug till time @seanlip @Showtim3?

bansalnitish commented:
Please resolve the merge conflicts.

seanlip commented:
> Even I'm not sure about dropping that todo -- definitely need some more time and investigation. Can we file a bug till time @seanlip @Showtim3?

@bansalnitish -- I'm suggesting doing some investigation so that you can add some more detail to the issue to determine what analysis has been done so far. At the moment there is no information in the issue that @Showtim3 filed to explain what, if anything, has been tried and why this is hard. Please see my [previous comment](https://github.com/oppia/oppia/pull/7818#discussion_r344017599) about this as well.

bansalnitish commented:
Okay sounds good -- we will dive in more and get back to you :).

Showtim3 committed:
Upgrading httpClient

Showtim3 committed:
Merging develop

Showtim3 committed:
Fixing imports

Showtim3 committed:
fixing upgraded services

Showtim3 commented:
Hey Guys, @seanlip @bansalnitish @vojtechjelinek @Jamesjay4199 

I did some research and found out that the circular dependency issue was present in angularjs http module(some $compile stuff) and that's why **$.ajax** was being used to make the http call instead of the http module of angularjs. But that isn't the case with angular, so with some brainstorming, I came to the conclusion that Jquery is not required in this case anymore and hence the ts-ignore is also not required anymore.

So i removed the jquery from this file and made the http call with the **HttpClient** of Angular and updated the spec file accordingly.

So, @vojtechjelinek @Jamesjay4199 I have made some significant changes, can you take another look?

@nitish i have introduced a variable **HttpClientDependency** which can be used for all the api services in the future.

Thanks!

vojtechjelinek commented on core/templates/dev/head/services/UpgradedServices.ts:
```diff
@@ -120,7 +124,13 @@ import { WrittenTranslationsObjectFactory } from
+          handle(req: HttpRequest<any>): Observable<HttpEvent<any>> {
+            return undefined;
+          }
+        });
```
Could you please explain this part? (add comment in code explaining it).

vojtechjelinek commented on core/templates/dev/head/services/csrf-token.service.ts:
```diff
@@ -16,40 +16,35 @@
+      throw new Error('Token request has already been made');
+    }
+
+    this.tokenPromise = this.httpClient.get('/csrfhandler').toPromise()
```
This is actually much cleaner than the jQuery solution! Good job!

vojtechjelinek commented on core/templates/dev/head/services/csrf-token.service.ts:
```diff
@@ -16,40 +16,35 @@
         return response.token;
       });
-    },
+  }
```
Could we do this in the service constructor? (I'm not sure if that is the ideal solution and if it would actually work)

vojtechjelinek commented on core/templates/dev/head/services/csrf-token.service.ts:
```diff
@@ -16,40 +16,35 @@
+    }
+
+    this.tokenPromise = this.httpClient.get('/csrfhandler').toPromise()
+      .then((response: any) => {
```
Could the `.then(` part be moved to the previous line?

vojtechjelinek reviewed (changes requested):
Thanks, good job! Just a few comments.

seanlip commented:
Hi @Showtim3, all the e2e tests are failing. Could you PTAL? Make sure to manually test on your computer (with the local dev server) before submitting a PR.

In general I think these changes are going in the correct direction (great job!) but please make sure that you don't break anything. That's really important!

Showtim3 committed:
Merge branch 'develop' into migration-CsrfTokenService

Showtim3 committed:
Reverting to jQuery

bansalnitish commented:
Travis tests are failing

Showtim3 commented:
@bansalnitish Was just a glitch..

bansalnitish commented:
Got it.

bansalnitish commented:
This PR uses `$ajax` to send request which gives typescript error.

`$.ajax()` thing can be removed using HttpClient via `import { HttpClientModule } from '@angular/common/http` (refer: https://angular.io/guide/http#setup). 

Since, we are in hybrid state and we need to create an instance of `CsrfTokenService` in `upgradedServices.ts` which will take an instance of HttpClient if we use it. We will need something `new CsrfTokenService(new HttpClient())`. 

We cannot create instance of HttpClient -- at least I've not seen creating instance of HttpClient using `new` rather than injecting the service. We can inject HttpClient through angular dependency injection in the constructor but I haven't found a way to create an instance manually apart from injecting.

After migration, we won't need to create instances using `new` -- the upgraded services exists as long as angular migration project exists. It would be that time wherein we can replace `$.ajax` with `this.HttpClient`. For now, we can have a TODO since there are services that depend on HttpClient.

Let me know your thoughts.

cc/ @seanlip @ankita240796 @Showtim3 @kevinlee12

seanlip commented:
Thanks @bansalnitish. Could you state the timeline more clearly? It sounds like "updating $.ajax" is blocked on "introducing HttpClient" but I am confused about what "introducing HttpClient" is blocked on.

We should have a plan of action which includes, somewhere, "introducing HttpClient" and "updating $.ajax" and it should be clear what the unblocker for each of the items in the plan is. This information should be present in any issues that are filed so that it's clear when they can be taken up.

bansalnitish commented:
Sorry for not writing it clearly.

Introducing HttpClient cannot be done until the angular migration wraps up as per my knowledge and is due to the fact that we cannot create an instance of HttpClient apart from injecting the HttpClient into the constructor -- like this `new CsrfTokenService(new HttpClient())` cannot be done(there is some abstract class that HttpClient uses further).

Once the angular migration is done, we will no longer need to provide the upgraded services as we do now to the AngularJS services by creating instances, hence we won't need the above syntax.

Let me know if it needs more clarity.

bansalnitish commented:
Umm -- one idea that came to my head. @Showtim3 can't we inject the HttpClient into the constructor of `UpgradedServices` like we do in services and then can create `CsrfService` instance writing something like `new CsrfService(this.httpClient)` ?

Showtim3 commented:
@bansalnitish I have already iterated over this idea...we can surely do what you are saying...but then everywhere that we are using the new Upgradedservices(), we'll have to provide new HttpClient() which is exactly the problem that we are facing right now. like `new UpgradedServices(new HttpClient())`. Is this clear?

bansalnitish commented:
Oh yes, you're right -- we will face the similar problem there too.

seanlip commented:
My [previous comment](https://github.com/oppia/oppia/pull/7818#issuecomment-559460172) still holds -- please file an issue stating in detail what needs to be done, what it's blocked on, and why it can't be done any earlier. Link to that issue here (in the comment thread), as well as in the TODO in the PR. The TODO in the PR should state clearly what needs to be done and what the prerequisites are for doing it.

Showtim3 commented:
@vojtechjelinek Hey we have to revert back to jQuery temporarily. PTAL! Thanks

vojtechjelinek commented on core/templates/dev/head/services/csrf-token.service.ts:
```diff
@@ -19,37 +19,45 @@
+
+    // We use jQuery here instead of Angular's $http, since the latter creates
+    // a circular dependency.
+    this.tokenPromise = $.ajax({
```
Should we add a TODO comment that jQuery should later be replaced?

Showtim3 replied:
@vojtechjelinek Done!

vojtechjelinek reviewed (changes requested):
LGTM! Just one question.

Showtim3 committed:
Merge branch 'develop' into migration-CsrfTokenService

Showtim3 committed:
adding a todo

Showtim3 commented:
@vojtechjelinek Added a todo #8035 . @seanlip PTAL!
Thanks

seanlip commented:
@Showtim3 The issue you've linked to says "not needed anymore".

As mentioned above -- please file an issue stating in detail what needs to be done, what it's blocked on, and why it can't be done any earlier. The issue linked to is not entirely clear on the last part.

Showtim3 commented:
@seanlip I have updated the linked issue...let me know if it's still unclear. PTAL! Thanks.

seanlip reviewed (approved):
Much better. Thanks @Showtim3! Please do this sort of thing in the future for PRs, so that we avoid unnecessary back-and-forth during reviews.

vojtechjelinek reviewed (approved):
LGTM!
